filestorage.js

```diff
@@ -0,0 +1,171 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const crypto = require('crypto');
+const common = require('metarhia-common');
+const { crc32 } = require('crc');
+const { zip, gzip } = require('compressing');
+
+const FS_EXT = 'f';
+
+class FileStorage {
+
+  // Create new FileStorage
+  //   dir - <string>, data storage directory
+  constructor(dir) {
+    this.dir = path.resolve(dir);
+    this.compressionList = [];
+    if (!fs.existsSync(dir)) fs.mkdirSync(dir);
+  }
+
+  _computeHash(data, checksum) {
+    switch (checksum) {
+      case 'CRC32':
+        return crc32(data).toString(16);
+      case 'SHA256':
+        return crypto.createHash('sha256')
+          .update(data)
+          .digest('hex');
+      default:
+        return '';
+    }
+  }
+
+  _mkdir(dir, cb) {
+    fs.access(dir, fs.constants.F_OK, err => {
+      if (err) {
+        this._mkdir(path.dirname(dir), err => {
+          if (err) cb(err);
+          else fs.mkdir(dir, cb);
+        });
+      } else {
+        cb(null);
+      }
+    });
+  }
+
+  _prepareFilepath(id, cb) {
+    const chunks = common.idToChunks(id);
+    const file = path.join(this.dir, chunks.join('/') + '.' + FS_EXT);
+    this._mkdir(path.dirname(file), err => {
+      cb(err, file);
+    });
+  }
+
+  // Write file to storage
+  //   id - <Object>, common.Uint64, id of file
+  //   data - <string> | <Buffer> | <Uint8Array>, data to be written
+  //   opts - <Object>
+  //     checksum - <string>, checksum type
+  //     dedupHash - <string>, second checksum type
+  //   cb - <Function>, callback
+  //     err - <error>
+  //     checksum - <string>, data checksum
+  //     dedupHash - <string>, second data checksum
+  //     originalSize - <number>, data size
+  write(id, data, opts = {}, cb) {
+    const checksum = this._computeHash(data, opts.checksum || 'CRC32');
+    if (checksum === '') {
+      cb(new Error(`${opts.checksum} is not supported`));
+      return;
```

Comment:
I think you can just `throw` in the default branch of _computeHash.
There is no need to have graceful errors (in cb) in case of unsupported operations, users shouldn't call this function with those arguments at all.
Also, this may even qualify for TypeError, though I'm not sure.